python_modules/libraries/dagster-components/dagster_components/core/schema/base.py

```diff
@@ -1,42 +1,118 @@
-from abc import abstractmethod
 from collections.abc import Mapping
-from typing import Any, Generic, TypeVar
+from typing import TYPE_CHECKING, Any, Callable, ClassVar, Generic, Optional, TypeVar
 
-from pydantic import BaseModel, ConfigDict, TypeAdapter
+from dagster._record import record
+from pydantic import BaseModel, ConfigDict
 
-from dagster_components.core.schema.metadata import get_resolution_metadata
-from dagster_components.core.schema.resolver import TemplatedValueResolver
+if TYPE_CHECKING:
+    from dagster_components.core.schema.context import ResolutionContext
+
+
+T_ResolverType = TypeVar("T_ResolverType", bound=type["Resolver"])
+T_ResolvableModel = TypeVar("T_ResolvableModel", bound="ResolvableModel")
 
 T = TypeVar("T")
+T_ResolveAs = TypeVar("T_ResolveAs")
+
+
+@record
+class _ResolverData:
+    """Container for configuration of a Resolver that is set when using the @resolver decorator."""
+
+    resolved_type: Optional[type]
+    renamed_fields: Mapping[str, str]
+
+
+class Resolver(Generic[T_ResolvableModel]):
+    """A Resolver is a class that can convert data contained within a ResolvableModel into an
+    arbitrary output type.
+
+    Methods on the Resolver class should be named `resolve_{fieldname}` and should return the
+    resolved value for that field.
+
+
+    Usage:
+
+        .. code-block:: python
+
+            class MyModel(ResolvableModel):
+                str_val: str
+                int_val: int
+
+            class TargetType:
+                def __init__(self, str_val: str, int_val_doubled: int): ...
+
+            @resolver(fromtype=MyModel, totype=TargetType, renamed_fields={"int_val": "int_val_doubled"})
```

Comment:
we can extend it later, but almost certainly we will want the ability to point to an arbitrary function that produces the `totype`